lock/postgres.go

```diff
@@ -0,0 +1,110 @@
+package lock
+
+import (
+	"context"
+	"database/sql"
+	"errors"
+	"fmt"
+	"time"
+
+	"github.com/sethvargo/go-retry"
+)
+
+// NewPostgresSessionLocker returns a SessionLocker that utilizes PostgreSQL's exclusive
+// session-level advisory lock mechanism.
+//
+// This function creates a SessionLocker that can be used to acquire and release locks for
+// synchronization purposes. The lock acquisition is retried until it is successfully acquired or
+// until the maximum duration is reached. The default lock duration is set to 60 minutes, and the
+// default unlock duration is set to 1 minute.
+//
+// See [SessionLockerOption] for options that can be used to configure the SessionLocker.
+func NewPostgresSessionLocker(opts ...SessionLockerOption) (SessionLocker, error) {
+	cfg := sessionLockerConfig{
+		lockID:        DefaultLockID,
+		lockTimeout:   DefaultLockTimeout,
+		unlockTimeout: DefaultUnlockTimeout,
+	}
+	for _, opt := range opts {
+		if err := opt.apply(&cfg); err != nil {
+			return nil, err
+		}
+	}
+	return &postgresSessionLocker{
+		lockID: cfg.lockID,
+		retryLock: retry.WithMaxDuration(
+			cfg.lockTimeout,
+			retry.NewConstant(2*time.Second),
+		),
+		retryUnlock: retry.WithMaxDuration(
+			cfg.unlockTimeout,
+			retry.NewConstant(2*time.Second),
+		),
+	}, nil
+}
+
+type postgresSessionLocker struct {
+	lockID      int64
+	retryLock   retry.Backoff
+	retryUnlock retry.Backoff
+}
+
+var _ SessionLocker = (*postgresSessionLocker)(nil)
+
+func (l *postgresSessionLocker) SessionLock(ctx context.Context, conn *sql.Conn) error {
+	return retry.Do(ctx, l.retryLock, func(ctx context.Context) error {
+		row := conn.QueryRowContext(ctx, "SELECT pg_try_advisory_lock($1)", l.lockID)
```

Comment:
It's possible someone may want to opt out of retries and prefer a blocking operation like `pg_advisory_lock`

> If another session already holds a lock on the same resource identifier, this function (_pg_advisory_lock_) will wait until the resource becomes available.

I suppose with an option like `lock.WithBlocking` we could extend and add this logic. While keeping the default as `pg_try_advisory_lock` with retries.